Source/DafnyStandardLibraries/examples/ConcurrentExamples.dfy

```diff
@@ -0,0 +1,177 @@
+module ConcurrentExamples {
+
+  import opened DafnyStdLibs.Concurrent
+  import opened DafnyStdLibs.Wrappers
+  import opened Helpers
+
+  const p1: nat -> bool := _ => true
+  const p2: (nat, nat) -> bool := (_, _) => true
+
+  datatype Copy = A | B
+
+  class Application {
+    var mutex: Lock
+    var box: AtomicBox<nat>
+    var primary: MutableMap<nat, nat>
+    var secondaryA: MutableMap<nat, nat>
+    var secondaryB: MutableMap<nat, nat>
+
+    constructor ()
+      ensures Valid()
+      ensures fresh({mutex, box, primary, secondaryA, secondaryB})
+    {
+      mutex := new Lock();
+      box := new AtomicBox(p1, 0);
+      primary := new MutableMap(p2);
+      secondaryA := new MutableMap(p2);
+      secondaryB := new MutableMap(p2);
+    }
+
+    ghost predicate Valid()
+      reads this`box, this`primary, this`secondaryA, this`secondaryB, this.box, this.primary, this.secondaryA, this.secondaryB
+    {
+      box.inv == p1 && primary.inv == p2 && secondaryA.inv == p2 && secondaryB.inv == p2
+      && box.Valid() && primary.Valid() && secondaryA.Valid() && secondaryB.Valid()
+    }
+
+    method Write(n: nat)
+      requires Valid()
+      modifies box
+      ensures Valid()
+    {
+      box.Put(n);
+    }
+
+    method Commit(slot: nat)
+      requires Valid()
+      modifies primary
+      ensures Valid()
+    {
+      var value := box.Get();
+      primary.Put(slot, value);
+    }
+
+    method Propagate(slot: nat)
+      requires Valid()
+      modifies secondaryA, secondaryB
+      ensures Valid()
+    {
+      mutex.Lock();
+      var val := primary.Get(slot);
+      if val.Some? {
+        secondaryA.Put(slot, val.value);
+      }
+      val := primary.Get(slot);
+      if val.Some? {
+        secondaryB.Put(slot, val.value);
+      }
+      mutex.Unlock();
+    }
+
+    method Read(copy: Copy, slot: nat) returns (r: Option<nat>)
+      requires Valid()
+    {
+      match copy
+      case A => r := secondaryA.Get(slot);
+      case B => r := secondaryB.Get(slot);
+    }
+
+    method EverythingElse()
+      requires Valid()
+      modifies primary
+      ensures Valid()
+    {
+      assert primary.inv(0,0);
+      primary.Remove(0);
+      var keys := primary.Keys();
+      print keys, "\n";
```

Comment:
We should replace the print statements with expects, since nothing looks at the output in this project.